### PR TITLE
Introduce Loadables and ExpandedPools

### PR DESCRIPTION
**Pull request by hammeiam**
Loadables adhere to the type 
```
export interface LoadableType<T> {
  isLoading: boolean
  isLoaded: boolean
  isError: boolean
  isSuccess: boolean
  error: Error | string | null
  data?: T
}
```

ExpandedPools are just like BasicPools except they are guaranteed to have all data about their tokens and underling tokens available
```
export type ExpandedPool = Omit<
  BasicPool,
  "tokens" | "underlyingTokens" | "lpToken"
> & {
  tokens: BasicToken[]
  underlyingTokens: BasicToken[]
  lpToken: BasicToken
}
```

### CODE REVIEW

**Review by ug02fast — APPROVED** [review]

File: src/providers/ExpandedPoolsProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
+import { BasicPool, BasicPoolsContext } from "./BasicPoolsProvider"
+import { BasicToken, TokensContext } from "./TokensProvider"
+import {
+  EMPTY_LOADABLE,
+  LoadableType,
+  useLoadingState,
+} from "../utils/loadable"
+import React, { ReactElement, useContext, useEffect } from "react"
+
+export type ExpandedPool = Omit<
```
Comment (nguyenaiden) [inline src/providers/ExpandedPoolsProvider.tsx:10]: For this, you're omitting the `"tokens" | "underlyingTokens" | "lpToken"` keys and readding it after. Could we potentially just overwrite here?

File: src/providers/ExpandedPoolsProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    },
+  )
+
+  useEffect(() => {
+    onStart()
+  }, [onStart])
+
+  useEffect(() => {
+    const errorPools = new Set<string>()
+    let errorTokensAll = new Set<string>()
```
Comment (nguyenaiden) [inline src/providers/ExpandedPoolsProvider.tsx:48]: Up to your discretion. Not a fan of reassignments, so I think we can do something like this:
```useEffect(() => {
    const errorPools: string[] = []
    const errorTokensAll: string[] = []
    if (basicPools && tokens) {
      const result = Object.values(basicPools || {}).reduce(
        (acc, basicPool) => {
          const errorTokens: string[] = []
          // Tokens
          const expandedPoolTokens = basicPool.tokens.map((tokenAddr) => {
            const token = tokens[tokenAddr]
            if (!token) {
              errorTokens.push(tokenAddr)
            }
            return token
          }) as BasicToken[]
          // Underlying Tokens
          const expandedUnderlyingPoolTokens = basicPool.isMetaSwap
            ? (basicPool.underlyingTokens.map((tokenAddr) => {
                const token = tokens[tokenAddr]
                if (!token) {
                  errorTokens.push(tokenAddr)
                }
                return token
              }) as BasicToken[])
            : null
          // LP Token
          const expandedLpToken = tokens[basicPool.lpToken]
          if (!expandedLpToken) {
            errorTokens.push(basicPool.lpToken)
          }

          // If any tokens aren't found, invalidate the whole pool
          if (errorTokens.length > 0) {
            errorTokensAll.concat(errorTokens)
            errorPools.push(basicPool.poolName)
            return acc
          }
          const expandedPool = {
            ...basicPool,
            tokens: expandedPoolTokens,
            underlyingTokens: expandedUnderlyingPoolTokens,
            lpToken: expandedLpToken,
          } as ExpandedPool
          return {
            byName: {
              ...acc.byName,
              [basicPool.poolName]: expandedPool,
            },
            byAddress: {
              ...acc.byAddress,
              [basicPool.poolAddress]: expandedPool,
            },
          }
        },
        { byName: {}, byAddress: {} },
      )

      if (errorTokensAll.length > 0) {
        const tokensString = [...new Set(errorTokensAll.values())].join(", ")
        const poolsString = [...new Set(errorPools.values())].join(", ")
        console.error(
          `Error loading tokens [${tokensString}] in pools [${poolsString}]`,
        )
      }

      onSuccess(result)
    }
  }, [basicPools, onSuccess, tokens])

**Review by nguyenaiden — APPROVED** [review]
2 comments, but LGTM!

File: src/providers/ExpandedPoolsProvider.tsx (hunk 1 above)
Comment (hammeiam) [reply src/providers/ExpandedPoolsProvider.tsx:10]: I'm not sure! Makes sense that we could though. I'll give it a shot.

File: src/providers/ExpandedPoolsProvider.tsx (hunk 2 above)
Comment (hammeiam) [reply src/providers/ExpandedPoolsProvider.tsx:48]: You're right, I like this better. Get those `let`s outta here.

File: src/providers/ExpandedPoolsProvider.tsx (hunk 1 above)
Comment (hammeiam) [reply src/providers/ExpandedPoolsProvider.tsx]: nope didn't work

**Pull request merged**